### DISCUSSION TITLE
Accessing values from the Posterior the right way

### DISCUSSION THREAD

**Discussion by PK1706** [pymc-labs/pymc-marketing#937] [Q&A]
Hi everyone,
I have some questions regarding the right way of accessing values from the posterior, depending on what I want to compare.

Based on PyMC-Marketing I implemented an MMM (some business requirements and conforming to legacy code made this necessary) which has the following structure:
![image](https://github.com/user-attachments/assets/deafbfab-b570-443e-80ad-d628f191c60d)

Part of the control variables are a **trend, holiday and seasonality component** which were created **using Prophet** (therefore I didn't use the Fourier contributions provided by PyMC-Marketing).
I used a Truncated Normal Distribution for the control variables to allow for negative coefficients for all control coefficients except for the ones related to Prophet.

To make this more structured I will number my questions:

1. To get **control contributions and channel contributions** I currently calculate the median (also tried it with the mean) of all of the parameters and then perform transformations and multiply by the coefficients. I do this in the same way as e.g. in the budget allocation code. When I on the other hand access the posterior median of channel or control contributions directly (also sometimes used in the PyMC-Marketing Code), this results in different values. What could be a possible cause of this and what is the right way to access these values?

2. What is the difference between extracting the **posterior median of mu and the posterior median of y**? Basically what I am confused about is when interpreting the influence of variables we basically only look at their influence on mu and not on the likelihood y right? What implications does this have, for example also when sampling from the posterior predictive distribution?

3. This question may be or may be not connected to the previous questions. Below you see...:
    a) the **posterior predictive plot** (which looks quite good)
  
    ![image](https://github.com/user-attachments/assets/9a4c0875-c692-44e7-a9c2-fb9fbec03f28)

    b) a **residual dependence plot** (comparing the observed values with the sum of channel contributions, control contributions and intercept) --> the dashed line is at 0
    
    ![image](https://github.com/user-attachments/assets/c49d627c-ad45-4d34-a54e-2e2d7cdaaf42)

   c) a goodness of fit plot (comparing the **observed values (red)** with the sum of median channel contributions, control contributions and intercept **(blue)**)
  
     ![image](https://github.com/user-attachments/assets/e0079dd6-8dad-456c-bd14-f123e1c03e54)

   **Unfortunately I had to remove the axis labels.**
   What is however directly noticeable is a kind of **constant bias especially visible in the residual dependence plot**. It almost looks like the model is trying to predict my observed value, plus a constant. In the Posterior predictive plot this is however not as visible.

   Do you have any idea what could lead to this bias? I have tried different kinds of priors, managing to reduce the bias (to the level you see in the plots) but not completely remove it. Am I accessing the wrong thing by using the median of the sum of median channel contributions, control contributions and intercept? If so, why would this be (vastly) different from the median of y and lead to this constant bias?

I hope that I could somehow convey what my problem is. If you need some more information to answer any of these questions please let me know!
I would highly appreciate an answer to any of these questions!

**Comment by williambdean** [comment]
Hi @PK1706 

Lots of items here so hopefully I can address at least a few:

1. Bit hard to judge this without an example or knowing what you are doing. Accessing the "control_contributions" and "channel_contributions" directly from the InferenceData would be the way to go. They already back in the uncertainty of the posterior parameters. They will have to be scaled back based on the max value in training y. 
2. `mu` in the graph is the mean estimate of the outcome variable whereas `y` in the graph is the outcome itself. This is the difference between mean vs an observation. Think confidence of average vs confidence of individual observation
If you take the mean across draws, chains, etc, then they will be close since the likelihood in this case is symmetric around mu. The variation of `y` will be greater than variation of `mu` since `y` has additional noise from Normal

**Reply by PK1706** [reply]
Thanks a lot for the quick reply @wd60622 ! Didn't expect an answer this quickly :)
You response already led me to uncover a bug, leading to the problem in 3:
![image](https://github.com/user-attachments/assets/4e6032cc-2436-4617-a411-3dea7b972e82)

In this plot you see adstocked spend on the x axis and contribution for one channel on the y axis. The red line is constructed using the median values of all the parameters and calculating the response "by hand". The blue line is the median of the channel_contributions posterior for this channel. Obviously there seems to be something wrong with the way I am calculating this judging from the constant bias. What exactly is wrong I have yet to uncover.

However the reason why I initially chose this method instead of accessing the channel_contributions "directly" was that there are some points in the blue curve which show non monotonic behavior (I highlighted 2 of them in black on the left side).
**Can you explain how these values come about? Given the model structure it doesn't make sense to me that a higher adstocked spend could lead to lower response** (of course there could still be diminishing returns, it should however still be higher/equal)

2. Ok thank you, so basically y adds an additional error term / uncertainty to mu (given a symmetric likelihood)?

**Reply by PK1706** [reply]
@wd60622 
I dug a little bit deeper and was able to identify the cause of the bug I was experiencing.

When calculating the median of all parameters and performing transformations with them (as shown in the following code block) this will lead to results that are off from the "channel_contributions" of the model.
```python
l_max = 8 # same as defined in the model
alpha = model.fit_result["alpha"].median(dim=["chain", "draw"]
lam = model.fit_result["lam"].median(dim=["chain", "draw"]
coefficients = model.fit_result["beta_channel"].median(dim=["chain", "draw"]

adstocked_spend = geometric_adstock(
  model.idata.constant_data["channel_data"].values, 
  alpha = alpha, 
  l_max = l_max, 
  normalize = True
)

saturated_spend = logistic_saturation(
  adstocked_spend,
  lam = lam
)

channel_contributions = saturated_spend * coefficients

# the channel_contributions which were calculated by hand differ from these channel_contributions_model which are directly extracted from the model
channel_contributions_model = model.fit_result["channel_contributions"].median(dim=["chain", "draw"]

```


When however performing transformations based on all chain/draw samples and only aggregating the results using the median in the end, I get the same results as the model:
```python
data = model.idata.constant_data["channel_data"].to_numpy()

alpha = model.fit_result["alpha"].to_numpy()
data_adstocked = geometric_adstock(data, alpha=alpha, l_max=8, normalize=True).eval()

lam_expanded= np.repeat(model.fit_result.lam.to_numpy()[:, :, np.newaxis, :], 156, axis=2)
data_saturated = logistic_saturation(data_adstocked, lam_expanded).eval()

data_saturated_x_array = xr.DataArray(
    data_saturated,
    dims=["chain", "draw", "date", "channel"],
    coords={
        "chain": np.arange(4),
        "draw": np.arange(2000),
        "date": model.idata.constant_data.date.values,
        "channel": model.idata.constant_data.channel.values
    }
)

coefficients = model.fit_result.beta_channel

channel_contributions = (data_saturated_x_array * coefficients).median(dim=["chain", "draw"])

# this way I get the same channel contributions as when I directly extract them from the model:
channel_contributions_model = model.fit_result["channel_contributions"].median(dim=["chain", "draw"]

```

Why does it matter at which step the posterior is aggregated, could it be due to limited floating point accuracy and small errors adding up?
If this is the case the response curves in the budget optimization would be erroneous as they are based on parameter medians. Do you have an idea how this could be mitigated or how this mismatch should be handled??

**Reply by cluhmann** [reply]
There is no error here. This is just how summary statistics (e.g., the median) and non-linear transformations work.   Compare the following:

```python
import numpy as np
mu, sigma = 0, 1
rng = np.random.default_rng()
draws = rng.normal(mu, sigma, 1000)

# take the square of the median
print(np.median(draws)**2)
# take the median of the square
print(np.median(draws**2))
```

Chances are, you want to delay collapsing your full posterior down to a point estimate (e.g., by taking the median) for as long as possible, only doing so when absolutely necessary. That's "the Bayesian way" of doing things.

**Reply by williambdean** [reply]
Great example above to recreate what is happening. In many scenarios, f(g(x)) != g(f(x)). In our case, g = media transformation and f = median
Here is related theorem at play: https://en.wikipedia.org/wiki/Jensen%27s_inequality

**Reply by PK1706** [reply]
@cluhmann 
The non-monotiny of the quadratic transformation function in your example leads to the deviation of results. However in our case we use a logistic saturation function, which is strictly monotonous and thus should commute with taking the median.
 
Can you think of any reason why your argument still appllies or do you have any other idea on what could be the issue ?

**Reply by williambdean** [reply]
I am struggling to understand why the PyMC internals cannot be relied upon and that the calculations have to be done again after already defined in the model graph.

I would trust the model graph over doing it manually. Is that required?

**Reply by PK1706** [reply]
@wd60622 I get where you are coming from.
The reason for calculating it manually came from the fact that when generating response curves directly from the model graph  there are non-monotonic patterns which I do not understand. When calculating the response manually this problem vanishes.
An example of this can be seen in my previous response:
> ![image](https://private-user-images.githubusercontent.com/74117444/358587367-4e6032cc-2436-4617-a411-3dea7b972e82.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjY0NzIxMDksIm5iZiI6MTcyNjQ3MTgwOSwicGF0aCI6Ii83NDExNzQ0NC8zNTg1ODczNjctNGU2MDMyY2MtMjQzNi00NjE3LWE0MTEtM2RlYTdiOTcyZTgyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MTYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTE2VDA3MzAwOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWZkMGNmOTAxYWQ0NTg1MmMzNWI3NTg4Y2I1Y2QxZGIxZDMzYmNlNGY3Njc4Yzc0OGE3OTEwYzVkNDM2MjMxNmEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.aAdyjOvVdj8irazoxgjB0mb8q1xaFRWf2Tl8nI6QDJU)
> 
> In this plot you see adstocked spend on the x axis and contribution for one channel on the y axis. The red line is constructed using the median values of all the parameters and calculating the response "by hand". The blue line is the median of the channel_contributions posterior for this channel. Obviously there seems to be something wrong with the way I am calculating this judging from the constant bias. What exactly is wrong I have yet to uncover.
> 
> **However the reason why I initially chose this method instead of accessing the channel_contributions "directly" was that there are some points in the blue curve which show non monotonic behavior (I highlighted 2 of them in black on the left side). Can you explain how these values come about? Given the model structure it doesn't make sense to me that a higher adstocked spend could lead to lower response (of course there could still be diminishing returns, it should however still be higher/equal)**

If you could provide an explanation for why at some levels of adstock spend an increase does not lead to a larger or at least the same response I would be more than happy to use the results from the model graph. Otherwise it is really hard to communicate the response curve to stakeholders.

I really want to thank you for all the ongoing support and valuable information, it is highly appreciated!

**Comment by cluhmann** [comment]
This has been sitting here for quite some time without a resolution. I believe that the issue stems from the the "by hand" calculations using the **marginal** medians/means of `alpha`, `lam`, and `coefficients`.  When dealing with the posterior distribution over model parameters $x$ and $y$, $f(E(x), E(y))$ will not, in general, be equivalent to $E(f(x, y))$, typically due to the (multivariate) geometry of the posterior (e.g., inferred dependencies among the values of parameters).  So the manual approach here is not calculating the same quantities (and is likely not calculating the quantities you want).
